library/src/main/java/io/ona/kujaku/location/clients/GoogleLocationClient.java

```diff
@@ -0,0 +1,298 @@
+import com.google.android.gms.location.LocationSettingsResult;
+import com.google.android.gms.location.LocationSettingsStatusCodes;
+import com.google.android.gms.tasks.OnSuccessListener;
+
+import io.ona.kujaku.R;
+import io.ona.kujaku.utils.LocationSettingsHelper;
+
+/**
+ * Created by Ephraim Kigamba - ekigamba@ona.io on 03/10/2018
+ */
+
+public class GoogleLocationClient extends BaseLocationClient implements LocationListener {
+
+    private Location lastLocation;
+    private FusedLocationProviderClient fusedLocationClient;
+    private GoogleLocationCallback googleLocationCallback;
+
+    private long updateInterval = 5000;
+    private long fastestUpdateInterval = 1000;
+
+    private static final String TAG = GoogleLocationClient.class.getName();
+    private static final int TWO_MINUTES = 1000 * 60 * 2;
+    private Object gpsStatusCallback;
+
+    public GoogleLocationClient(@NonNull Context context) {
+        this.context = context;
+        locationManager = (LocationManager) context.getApplicationContext().getSystemService(Context.LOCATION_SERVICE);
+        fusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
+        googleLocationCallback = new GoogleLocationCallback();
+    }
+
+    @Override
+    public void stopLocationUpdates() {
+        if (isMonitoringLocation()) {
+            lastLocation = null;
+            setLocationListener(null);
+            fusedLocationClient.removeLocationUpdates(googleLocationCallback);
+        }
+
+        unregisterForGpsStatusStop();
+    }
+
+    @Override
+    @Nullable
+    public Location getLastLocation() {
+        return lastLocation;
+    }
+
+    @Override
+    public void onLocationChanged(Location location) {
+        lastLocation = location;
+
+        if (getLocationListener() != null) {
+            getLocationListener().onLocationChanged(location);
+        }
+    }
+
+    @Override
+    public void requestLocationUpdates(@NonNull android.location.LocationListener locationListener) {
+        LocationRequest locationRequest = new LocationRequest();
+        locationRequest.setInterval(updateInterval);
+        locationRequest.setFastestInterval(fastestUpdateInterval);
+        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);
+        requestLocationUpdates(locationListener, locationRequest);
+    }
+
+    public void requestLocationUpdates(@NonNull android.location.LocationListener locationListener
+            , @NonNull LocationRequest locationRequest) {
+        setLocationListener(locationListener);
+        if (isProviderEnabled()) {
+            try {
+                fusedLocationClient.getLastLocation()
+                        .addOnSuccessListener((Activity) context, new OnSuccessListener<Location>() {
+                            @Override
+                            public void onSuccess(Location location) {
+                                if (lastLocation == null || location.getTime() > lastLocation.getTime()) {
```

Comment:
Why isn't this using the refined location accuracy method `isBetterLocation`?

Reply:
Could location be null here causing a NPE?

Reply:
Cool, fixing this